### ISSUE TITLE
The Impossible Problem

### ISSUE THREAD

**Issue by nrwiersma** [msales/streams#21]
**Note:** This issue should serve as a discussion about the problem below.

**Problem:**
Currently, when a stream is branched, and one sink commits, there is is a potential for data loss in the other branch. The source has committed, but we have an outstanding uncommitted batch. With this issue, it is only safe to use branches where all but one branch are sidecar branches (non-batching, non-committing)

The difficulty is that each node is running in a goroutine, forcing a commit safely and synchronously is difficult.

**Use Case:**
```
           d - e
          /
a - b - c
          \
           f - g
```
Both `e` and `g` are sinks that commit in batches. When `e` competes its batch and commits, `g` should be forced to commit its batch.

**Initial Idea:**
The initial idea has the following steps:
1. Add a Metadata Store that is tasked with assembling the current state of all sources, and the nodes that submitted the metadata.
2. Message Metadata should no longer be accumulative. Metadata can implement an interface indicating if it can be merged (MergableMetadata). If it is mergeable, it should be merged into the Metadata Store, otherwise it can be replaced. (This will also get rid of ugly code in MergeProcessor)
3. Pipe interface should implement a `Mark` function indicating the message was discarded. The pipe should submit *marks* to the Metadata Store. All messages should be marked, forwarded or committed.
4. A new interface Committer should be created with a `Commit` function. Sinks that can call commit should implement this interface. Sidecar sinks can opt out by not implementing the interface.
5. On committing, the current state should be gathered from the Metadata Store, and all nodes (involved with the current node, worked out by the node->source relationship) should be forced to commit **(this is the issue. No idea how to do this part)**.

**Sideline Benefits:**
1. We will have better handling of metadata, and this should cleanup some ugly code around this.
2. We can implement system-wise commits if no commit has been seen in a given time period. This will allow for cleaner, more naive streams.

**Comment by nrwiersma** [comment]
Initial design work can be found in https://github.com/msales/streams/tree/impossible

**Comment by nrwiersma** [comment]
Fixed in #23
